In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os
os.chdir('/content/drive/MyDrive/Geek for Geeks/Neural Network/Feedforward Neural Network')

In [4]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.losses import SparseCategoricalCrossentropy
from tensorflow.keras.metrics import SparseCategoricalAccuracy

**Load and prepare the MNIST dataset**

In [5]:
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0


11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


**Build the model**

In [6]:
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


**Compile the model**

In [7]:
model.compile(optimizer=Adam(),
              loss=SparseCategoricalCrossentropy(),
              metrics=[SparseCategoricalAccuracy()])

**Train the model**

In [8]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - loss: 0.4351 - sparse_categorical_accuracy: 0.8765
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - loss: 0.1237 - sparse_categorical_accuracy: 0.9641
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - loss: 0.0794 - sparse_categorical_accuracy: 0.9770
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - loss: 0.0555 - sparse_categorical_accuracy: 0.9830
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - loss: 0.0447 - sparse_categorical_accuracy: 0.9867


**Evaluate the model**

In [9]:
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f'\nTest accuracy: {test_acc}')

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 0.0911 - sparse_categorical_accuracy: 0.9742

Test accuracy: 0.9779000282287598


**Conclusion**

In this project, we built and trained a simple neural network using TensorFlow and Keras to classify handwritten digits from the MNIST dataset. The model architecture consisted of a Flatten layer to convert 2D image data into a 1D vector, followed by a Dense hidden layer with 128 neurons using the ReLU activation, and an output layer with 10 neurons using the softmax activation for multi-class classification.

After training for 5 epochs, the model achieved a training accuracy of approximately 98.6% and a test accuracy of about 97.8%, demonstrating strong generalization to unseen data. These results show that even a relatively simple feed-forward neural network can effectively recognize handwritten digits when trained on normalized image data.

This project highlights the effectiveness of neural networks for basic image classification tasks and provides a solid foundation for exploring more advanced architectures such as Convolutional Neural Networks (CNNs) to further improve accuracy and robustness.